packages/rpc-core/src/rpc-methods/getBlockProduction.ts

```diff
@@ -0,0 +1,30 @@
+import { Base58EncodedAddress } from '@solana/keys';
+import { Commitment, RpcResponse, U64UnsafeBeyond2Pow53Minus1 } from './common';
+
+type NumberOfLeaderSlots = number;
+type NumberOfBlocksProduced = number;
+
+type Range = Readonly<{
+    firstSlot: U64UnsafeBeyond2Pow53Minus1;
+    lastSlot: U64UnsafeBeyond2Pow53Minus1;
+}>;
+
+type GetBlockProductionApiResponse = RpcResponse<{
+    byIdentity: Readonly<{
+        [address: string]: [NumberOfLeaderSlots, NumberOfBlocksProduced];
+    }>;
+    range: Range;
+}>;
+
+export interface GetBlockProductionApi {
+    /**
+     * Returns recent block production information from the current or previous epoch.
+     */
+    getBlockProduction(
+        config?: Readonly<{
+            commitment?: Commitment;
+            identity?: Base58EncodedAddress;
+            range?: Range;
+        }>
+    ): GetBlockProductionApiResponse;
```

Comment:
Make overloads! You have two scenarios here which will have different return types.

* Has an identity => `byIdentity` is _definitely_ `{[TAddress in TIdentity]?: [NumberOfLeaderSlots, NumberOfBlocksProduced]}`
* Omits the identity => `byIdentity` is `Record<Base58EncodedAddress, [NumberOfLeaderSlots, NumberOfBlocksProduced]>`

```suggestion
    getBlockProduction(
        config?: Readonly<{
            commitment?: Commitment;
            identity: undefined
            range?: Range;
        }>
    ): GetBlockProductionApiResponseBase & GetBlockProductionApiResponseDefinitelyByIdentityOrNothing;
    getBlockProduction(
        config?: Readonly<{
            commitment?: Commitment;
            identity: Base58EncodedAddress;
            range?: Range;
        }>
    ): GetBlockProductionApiResponseBase;
```

…or whatever actually works, and think of names that aren't stupid. Look at `getAccountInfo` for some intense overrides.

Reply:
Nice! I've updated this to use overloads, definitely seems better!